Content.Server/Decals/DecalSystem.cs

```diff
@@ -412,46 +452,30 @@ public override void Update(float frameTime)
                 foreach (var (gridId, gridChunks) in chunksInRange)
                 {
                     var newChunks = _chunkIndexPool.Get();
-                    newChunks.UnionWith(gridChunks);
-                    if (_previousSentChunks[playerSession].TryGetValue(gridId, out var previousChunks))
-                    {
-                        newChunks.ExceptWith(previousChunks);
-                    }
+                    _dirtyChunks.TryGetValue(gridId, out var dirtyChunks);
 
-                    if (_dirtyChunks.TryGetValue(gridId, out var dirtyChunks))
+                    if (!previouslySent.TryGetValue(gridId, out var previousChunks))
+                        newChunks.UnionWith(gridChunks);
+                    else
                     {
-                        var inRange = new HashSet<Vector2i>();
-                        inRange.UnionWith(gridChunks);
-                        inRange.IntersectWith(dirtyChunks);
-                        newChunks.UnionWith(inRange);
+                        foreach (var index in gridChunks)
+                        {
+                            if (!previousChunks.Contains(index) || dirtyChunks != null && dirtyChunks.Contains(index))
+                                newChunks.Add(index);
+                        }
+
+                        previousChunks.Clear();
+                        _chunkIndexPool.Return(previousChunks);
                     }
 
+                    previouslySent[gridId] = gridChunks;
+
                     if (newChunks.Count == 0)
-                    {
                         _chunkIndexPool.Return(newChunks);
-                        continue;
-                    }
-
-                    // TODO: This is gonna have churn but mainly I want to fix the bugs rn.
-                    _previousSentChunks[playerSession][gridId] = gridChunks;
-                    updatedChunks[gridId] = newChunks;
```

Comment:
This was the actual bug, where `_previousSentChunks` was not being updated if the state didn't have any new chunks, despite some chunks possibly having been removed from the player's view.